### PR TITLE
Add CheckboxGroup and RadioGroup components

### PR DESCRIPTION
**Pull request by joshfarrant**
## Summary

Adds `CheckboxGroup` and `RadioGroup` components which, for the time being, are just implementations of the new private `ControlGroup` component. These components allow us to group `Radio` and `Checkbox` components using a `<fieldset>` and label them with a `<legend>`, providing a more accessible and semantic way to group related form elements.

## List of notable changes:

- Adds `ControlGroup` component (alternate name suggestions welcome)
- Adds `CheckboxGroup` and `RadioGroup` components, tests, and stories

## Still to do (in future PRs)

- Add documentation for the new components
- Consume new components in our existing documentation and stories

## What should reviewers focus on?

- There were some spacing inconsistencies between [the Figma file I was using for reference](https://www.figma.com/design/BJ95AjraesmRCWsKA013GS/Primer-Brand?node-id=2742-33835&t=uJa6xbnTmcheIEoo-4) and our actual form element sizes in Primer Brand.
  - Specifically, font sizes varied between the two. Figma has the font sizes of the Radio and Checkbox labels, as well as the validation, at 14px, but in PB they're all 16px. Could you advise which way to go with this please @danielguillan, as well as giving the design of the components a quick check too please?
- I used the new(ish) [CSF3](https://storybook.js.org/docs/api/csf#upgrading-from-csf-2-to-csf-3) in the stories. The only difference is the fact that stories now export objects with a render function, rather than exporting components. Although I didn't use it here, this has the added benefit of letting you create/update component stories from within the Storybook UI and save the new stories to disk. Happy to revert this if we'd prefer to keep things consistent, but at some point it might be a good idea to update all stories to CSF3 as one less thing to worry about for future Storybook upgrades.

## Steps to test:

1. Take a look at the Storybook 

## Supporting resources (related issues, external links, etc):

- https://github.com/github/primer/issues/3761

## Contributor checklist:

- [x] All new and existing CI checks pass
- [x] Tests prove that the feature works and covers both happy and unhappy paths
- [x] Any drop in coverage, breaking changes or regressions have been documented above
- [x] UI Changes contain new visual snapshots (generated by adding `update snapshots` label to the PR)
- [x] All developer debugging and non-functional logging has been removed
- [x] Related issues have been referenced in the PR description

## Reviewer checklist:

- [ ] Check that pull request and proposed changes adhere to our [contribution guidelines](../../CONTRIBUTING.md) and [code of conduct](../../CODE_OF_CONDUCT.md)
- [ ] Check that tests prove the feature works and covers both happy and unhappy paths
- [ ] Check that there aren't other open Pull Requests for the same update/change

## Screenshots:

> Please try to provide before and after screenshots or videos

<table>
<tr>
<th> RadioGroup </th> <th> CheckboxGroup </th>
</tr>
<tr>
<td valign="top">

image

 </td>
<td valign="top">

image

</td>
</tr>
</table>

### CODE REVIEW

File: .changeset/popular-seas-yawn.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react-brand': patch
+---
+
+Added new `CheckboxGroup` and `RadioGroup` components to semantically group related `Checkbox` and `Radio` components respectively.
```
Comment (rezrah) [inline .changeset/popular-seas-yawn.md:5]:
```suggestion
Added new `CheckboxGroup` and `RadioGroup` components to group `Checkbox` and `Radio` components respectively.
```

File: .changeset/popular-seas-yawn.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react-brand': patch
```
Comment (rezrah) [inline .changeset/popular-seas-yawn.md:2]:
```suggestion
'@primer/react-brand': minor
```

File: apps/docs/content/components/CheckboxGroup/index.mdx [hunk 3]
```diff
@@ -0,0 +1,49 @@
+---
+title: CheckboxGroup
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/index.mdx:2]: FYI we don't use PascalCase in docs unless we're referring to the React component in the React tab. Should use sentence case.

File: apps/docs/content/components/CheckboxGroup/index.mdx [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+description: Use CheckboxGroup to group related checkbox inputs with supporting content
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/index.mdx:3]: For consistency with its product counterpart

```suggestion
description: Checkbox group renders a set of checkboxes.
```

File: apps/docs/content/components/CheckboxGroup/index.mdx [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+---
+
+import ComponentLayout from '../../../src/layouts/component-layout'
+export default ComponentLayout
+
+## Usage
+
+The `CheckboxGroup` family of components provide a way to group related [Checkbox](/components/Checkbox) components together with supporting content such as labels, captions, and validation messages. It creates a semantic fieldset structure that helps maintain proper form hierarchy and accessibility.
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/index.mdx:11]: As it's already undergone various iterations, could we match the content from this page please? https://primer.style/components/checkbox-group

File: apps/docs/content/components/CheckboxGroup/react.mdx [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+import {Label} from '@primer/react'
+import {PropTableValues} from '../../../src/components'
+
+```js
+import {CheckboxGroup} from '@primer/react-brand'
+```
+
+## Examples
+
+### Basic
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/react.mdx:18]:
```suggestion
### Default
```

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+  padding: 0;
+  margin-bottom: var(--base-size-8);
+}
+
+.ControlGroup__caption {
+  font-size: var(--brand-text-size-100);
+  letter-spacing: var(--brand-text-letterSpacing-100);
+  line-height: var(--brand-text-lineHeight-100);
+  color: var(--brand-color-text-muted);
+}
```
Comment (rezrah) [inline packages/react/src/forms/ControlGroup/ControlGroup.module.css:24]: You should be able to get these styles for free by using <Text size="100" />. Can we use that here or are these styles doing something a bit different?

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+.ControlGroup__validation {
+  display: flex;
+  gap: var(--base-size-4);
+  font-size: var(--brand-text-size-100);
+  font-weight: var(--base-text-weight-regular);
+  line-height: var(--brand-text-lineHeight-100);
```
Comment (rezrah) [inline packages/react/src/forms/ControlGroup/ControlGroup.module.css:31]: This is quite a bit smaller than the usual FormControl validation, does it need to be? That one is 16px vs 14px here. Would recommend we keep that the same for consistency.

FormControl:
Screenshot 2024-11-27 at 13 43 01

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+  color: var(--brand-color-success-fg);
+}
+
+.ControlGroup__validation--error {
+  color: var(--brand-color-error-fg);
+}
+
+.ControlGroup__validation-icon--success {
+  position: relative;
+  top: -1.6px;
```
Comment (rezrah) [inline packages/react/src/forms/ControlGroup/ControlGroup.module.css:44]: Just curious why this is -1.6px vs -1px? I think -1px looks more vertically aligned to the center, but this rule is so specific I'm probably missing something:

With -1px:

Screenshot 2024-11-27 at 13 47 43

File: apps/docs/content/components/CheckboxGroup/react.mdx [hunk 10]
```diff
@@ -0,0 +1,185 @@
+---
+title: CheckboxGroup
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/react.mdx:2]:
```suggestion
Checkbox group
```

File: apps/docs/content/components/CheckboxGroup/react.mdx [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+| `children` | `string` |         | Caption text |
+
+`CheckboxGroup.Caption` extends the `span` element and supports all `span` props.
+
+### CheckboxGroup.Validation
+
+| Name       | Type                     | Default | Description                        |
+| :--------- | :----------------------- | :-----: | :--------------------------------- |
+| `children` | `string`                 |         | Validation message                 |
+| `variant`  | `'success'` \| `'error'` |         | Sets the validation state and icon |
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/react.mdx:183]: For visual consistency, can we use PropTableValues when showing more than one value please?

E.g.

```suggestion
| `variant` | <PropTableValues values={['error', 'success' ]} addLineBreaks /> | | Sets the validation state and icon |
```

File: .changeset/popular-seas-yawn.md (hunk 2 above)
Comment (joshfarrant) [reply .changeset/popular-seas-yawn.md:2]: Updated this, but just want to double-check my understanding.

I know we use minor for breaking changes, but do we also default to minor when adding a new component too?

File: apps/docs/content/components/CheckboxGroup/index.mdx (hunk 5 above)
Comment (joshfarrant) [reply apps/docs/content/components/CheckboxGroup/index.mdx:11]: Happy to make the change, but I just want to check before I do it as writing these docs was a decent chunk of work this morning 😅

Is the request to ditch what I've currently got and copy/paste the page from the PRC docs, updating any minor things that differ between the two implementations?

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+
+        {validationChildren ? (
+          <CheckboxGroup.Validation variant={validationVariant}>{validationChildren}</CheckboxGroup.Validation>
+        ) : null}
+      </CheckboxGroup>
+    )
+  },
+}
+
+export const Inline: Story = {
```
Comment (rezrah) [inline packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:99]: Could we move this story into a separate `CheckboxGroup.features.stories.tsx` file and organize it inside a `CheckboxGroup/Examples` or `CheckboxGroup/Features` subfolder please? Probably former in this case as it's not a feature of the component.

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+  args: {
+    labelChildren: 'Inline example',
+    labelVisuallyHidden: true,
```
Comment (rezrah) [inline packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:102]: Nit. to keep this example focussed on the inline aspect and discourage visually hiding labels where possible, I think we should ditch the caption and show the label.

File: .changeset/popular-seas-yawn.md (hunk 2 above)
Comment (rezrah) [reply .changeset/popular-seas-yawn.md:2]: > but do we also default to minor when adding a new component too?

👍 Yup breaking changes and new components...

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 7 above)
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:24]: Ah great, thanks. I think I just pulled these straight from [FormControl where they're hard-coded](https://github.com/primer/brand/blob/main/packages/react/src/forms/FormControl/FormControl.module.css#L71-L74).

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 9 above)
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:44]: Again, pulled from FormControl to keep things consistent

https://github.com/primer/brand/blob/main/packages/react/src/forms/FormControl/FormControl.module.css#L94-L97

Happy to modify it if you'd prefer though?

File: apps/docs/content/components/CheckboxGroup/index.mdx (hunk 5 above)
Comment (rezrah) [reply apps/docs/content/components/CheckboxGroup/index.mdx:11]: > Happy to make the change, but I just want to check before I do it as writing these docs was a decent chunk of work this morning 😅

Sorry @joshfarrant 😓 I should have mentioned this earlier. For future reference, you only add the React docs if when you're adding a new React component, the interface guidelines can land later unless it's easy enough to add in the first instance. Usually interface guidelines should include images too, but again can be added in a fast-follow PR if needed.

> Is the request to ditch what I've currently got and copy/paste the page from the PRC docs, updating any minor things that differ between the two implementations?

These two components are a bit unique in that the product counterpart already has mature documentation around this. Also given that our component shares the same API and usage guidelines as product and is essentially the same thing, my recommendation is to keep the content and structure of the guidelines as consistent as possible. You're fine to tweak aspects of the content to suit the unique aspects of our component though.

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 9 above)
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:44]: Ah, we've had this conversation before, just in the other direction 😄

It was 1px, but after a brief conversation ([here](https://github.com/primer/brand/pull/668#issuecomment-2250980453)) we settled on 1.6px.

https://github.com/primer/brand/pull/668/commits/4f50c91eea0331d01fe9157e81d1d6bd8434516d#diff-66f88db62ba93924cf6de6ae216861c98149f0122e02c93173b5778c43cd4b57R95-R98

Comment (rezrah) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:44]: Ah okay 🤣 .. let's keep it the same then.. i used a visual guide in the other screenshot so that's probably more correct. Although could this be a bit different because the text is 100 vs 200?

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 8 above)
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:31]: There's a bit more context on this in my other PR here which makes the change across the rest of PB.

https://github.com/primer/brand/pull/831

TL;DR is that our FormControl validation size didn't match Figma, so after chatting to Dani we've agreed to bring it down in the library to keep it consistent. The same goes for the hint.

File: packages/react/src/forms/ControlGroup/ControlGroup.tsx [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+      />
+    </ControlGroupContext.Provider>
+  )
+})
+
+export type ControlGroupLabelProps = {visuallyHidden?: boolean} & HTMLAttributes<HTMLLegendElement>
+const ControlGroupLabel = forwardRef<HTMLLegendElement, ControlGroupLabelProps>(
+  ({className, visuallyHidden, ...props}, ref) => {
+    return (
+      <legend
```
[the remark this answers is not in the document]
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.tsx:60]: Great spot, thanks!

Updated

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.test.tsx [hunk 15]
```diff
@@ -0,0 +1,153 @@
...
+        <FormControl>
+          <FormControl.Label>Choice three</FormControl.Label>
+          <Checkbox value="three" />
+        </FormControl>
+
+        <CheckboxGroup.Validation variant="success">Great job!</CheckboxGroup.Validation>
+      </CheckboxGroup>,
+    )
+
+    expect(getByRole('group', {name: /choices/i})).toBeInTheDocument()
```
[the remark this answers is not in the document]
Comment (joshfarrant) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.test.tsx:36]: No, just force of habit and, for whatever reason, I've seen the case-insensitive regex preferred in the past. Happy to switch to `{name: 'Choices'}` if you'd prefer though

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+      table: {
+        category: 'Validation',
+      },
+    },
+  },
+}
+
+export default meta
+
+type Story = StoryObj<MetaProps>
```
[the remark this answers is not in the document]
Comment (joshfarrant) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:69]: `StoryObj` is the built-in, recommended type. See the example in the [docs here](https://storybook.js.org/docs/writing-stories/typescript#typing-stories-with-meta-and-storyobj).

File: apps/docs/content/components/CheckboxGroup/index.mdx (hunk 5 above)
Comment (joshfarrant) [reply apps/docs/content/components/CheckboxGroup/index.mdx:11]: No that's all good, thank you for the additional context.

I'll tweak the interface docs to match PRC shortly. Are you happy with the react docs as they are, or shall I do the same there?

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.test.tsx (hunk 15 above)
[the remark this answers is not in the document]
Comment (joshfarrant) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.test.tsx]: I opted to remove the regex anyway

8405890e2e64c23a4934e6d0738591bd9ccc6b69

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 8 above)
Comment (rezrah) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css]: Gotcha, good to know 👍

File: packages/react/src/forms/RadioGroup/RadioGroup.test.tsx [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        <FormControl>
+          <FormControl.Label>Choice two</FormControl.Label>
+          <Radio name="demo" value="two" />
+        </FormControl>
+        <FormControl>
+          <FormControl.Label>Choice three</FormControl.Label>
+          <Radio name="demo" value="three" />
+        </FormControl>
+
+        <RadioGroup.Validation variant="success">Great job!</RadioGroup.Validation>
```
Comment (rezrah) [inline packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:32]: There's a Jest warning related to validation prop in the test logs. Possible to take a look at fixing that please?

https://github.com/primer/brand/actions/runs/12053705065/job/33609998572?pr=830#step:10:53

Screenshot 2024-11-27 at 16 59 11


(Doesn't seem to be happening for CheckboxGroup FWIW)

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 18]
```diff
@@ -0,0 +1,52 @@
+---
+title: Radio Group
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:2]:
```suggestion
title: Radio group
```

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 19]
```diff
@@ -0,0 +1,52 @@
...
+---
+
+import ComponentLayout from '../../../src/layouts/component-layout'
+export default ComponentLayout
+
+## Usage
+
+The `RadioGroup` family of components provide a way to group related [Radio](/components/Radio) components together with supporting content such as labels, captions, and validation messages. It creates a semantic fieldset structure that helps maintain proper form hierarchy and accessibility.
+
+> **Important**: When using Radio components within a RadioGroup, ensure you provide the same `name` prop to each Radio. This ensures they are correctly grouped and only one option can be selected at a time.
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:13]: > **Important**: When using Radio components within a radio group, ensure you provide the same `name` prop to each Radio. This ensures they are correctly grouped and only one option can be selected at a time.

File: apps/docs/content/components/CheckboxGroup/react.mdx [hunk 20]
```diff
@@ -0,0 +1,185 @@
...
+title: Checkbox group
+description: Checkbox group renders a set of checkboxes.
+source: https://github.com/primer/brand/blob/main/packages/react/src/forms/CheckboxGroup/CheckboxGroup.tsx
+figma: 'https://www.figma.com/design/BJ95AjraesmRCWsKA013GS/Primer-Brand?node-id=10467-3273&t=GA5GtEqPtQ9yeRaN-4'
+storybook: '/brand/storybook/?path=/story/components-forms-checkboxgroup--playground'
+---
+
+import {Label} from '@primer/react'
+import {PropTableValues} from '../../../src/components'
+
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/react.mdx:11]: Tabs aren't working in the React docs, making it hard to go back to overview. Think you're just missing this:

```suggestion

import ComponentLayout from '../../../src/layouts/component-layout'
export default ComponentLayout

```

Issue:


https://github.com/user-attachments/assets/afd35600-9ee8-45be-a394-6cfbf8f77e89

File: packages/react/src/forms/RadioGroup/RadioGroup.test.tsx (hunk 17 above)
Comment (joshfarrant) [reply packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:32]: Ah yea, thanks for raising this. I spotted this the other day and I meant to dig into it more.

It boils down to the fact that, in `FormControl.tsx`, we forward `validationStatus` to the Radio component, but Radio doesn't implement it, so it just ends up in the HTML.

I'm not sure if I can just stop FormControl from forwarding `validationStatus` to Radio, or whether Radio should actually be using it to set `aria-invalid` as Checkbox does.

My inclination is to tackle this in its own PR as it's not an issue introduced in this PR, wdyt?

File: apps/docs/content/components/RadioGroup/index.mdx (hunk 19 above)
Comment (joshfarrant) [reply apps/docs/content/components/RadioGroup/index.mdx]: I've just replaced this file with a (near) copy of PRC, lmk what you think

File: apps/docs/content/components/CheckboxGroup/index.mdx (hunk 5 above)
Comment (joshfarrant) [reply apps/docs/content/components/CheckboxGroup/index.mdx]: I've just replaced the interface guidelines with a (near) copy of PRC, lmk what you think

Comment (joshfarrant) [reply apps/docs/content/components/CheckboxGroup/index.mdx]: There's only the default example in the product React docs, so I think we're good on that front 👍

File: apps/docs/content/components/CheckboxGroup/react.mdx (hunk 20 above)
Comment (joshfarrant) [reply apps/docs/content/components/CheckboxGroup/react.mdx:11]: Well spotted, thank you

File: packages/react/src/forms/RadioGroup/RadioGroup.test.tsx (hunk 17 above)
Comment (joshfarrant) [reply packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:32]: Thinking about it though, I'm not entirely sure how an individual Radio control (not a group, just a single control) could ever be _invalid_ 🤔

Comment (joshfarrant) [reply packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:32]: I've removed the prop forwarding and resolved the `console.error` in the tests. Removing the forwarding hasn't modified the behaviour of FormControl or Radio in any way as the forwarded prop was never used.

8e910131f7366ebb0adc1ea3c1d3e9f262c52ef6

File: apps/docs/content/components/RadioGroup/react.mdx [hunk 21]
```diff
@@ -0,0 +1,164 @@
+---
+title: Radio Group
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/react.mdx:2]:
```suggestion
title: Radio group
```

File: apps/docs/content/components/CheckboxGroup/index.mdx [hunk 22]
```diff
@@ -0,0 +1,43 @@
...
+
+**Label:** A title that labels the group of options and helps users understand the relationship between the options in the group.
+
+**Required indicator:** Indicates that a selection is required.
+
+**Caption:** May be used to provide additional context about the checkbox group when the label alone is not sufficient
+
+**Options:** A list of options represented using [checkboxes](/components/Checkbox).
+
+**Validation message:** A message explaining why the selection failed validation. See the [form pattern documentation](/ui-patterns/forms/overview#validation) for more information on form validation patterns.
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/index.mdx:33]: This link 404's, so recommend we use an external link:

```suggestion
**Validation message:** A message explaining why the selection failed validation. See the [form pattern documentation](https://primer.style/ui-patterns/forms/overview#validation-message) for more information on form validation patterns.
```

File: apps/docs/content/components/CheckboxGroup/index.mdx [hunk 23]
```diff
@@ -0,0 +1,43 @@
...
+
+## Accessibility
+
+Once a user checks an option from a checkbox group, the result should not be saved or applied until the user has explicitly submitted their input using a "save" or "submit" button. See the [saving pattern guidelines](/ui-patterns/saving) for more information.
```
Comment (rezrah) [inline apps/docs/content/components/CheckboxGroup/index.mdx:37]: Same as above:

```suggestion
Once a user checks an option from a checkbox group, the result should not be saved or applied until the user has explicitly submitted their input using a "save" or "submit" button. See the [saving pattern guidelines](https://primer.style/ui-patterns/saving) for more information.
```

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+import anatomy from './images/anatomy.png'
+export default ComponentLayout
+
+## Usage
+
+Use radio group to allow users to select a single option from a short list of related options.
+
+### Orientation
+
+A vertical orientation makes options easier to visually scan. A horizontal orientation is not yet supported.
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:17]: Let's remove the last sentence as we provide an example in the next tab for it...

```suggestion
A vertical orientation makes options easier to visually scan.
```

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+
+**Label:** A title that labels the group of options and helps users understand the relationship between the options in the group
+
+**Required indicator:** Indicates that a selection is required
+
+**Caption:** May be used to provide additional context about the radio group when the label alone is not sufficient
+
+**Options:** A list of mutually exclusive options represented using [radio buttons](/components/Radio)
+
+**Validation message:** A message explaining why the selection failed validation. See the [form pattern documentation](/ui-patterns/forms/overview#validation) for more information on form validation patterns.
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:31]:
```suggestion
**Validation message:** A message explaining why the selection failed validation. See the [form pattern documentation](https://primer.style/ui-patterns/forms/overview#validation-message) for more information on form validation patterns.
```

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+- Put radio buttons in a logical order
+- Only use a radio group for a short list of options: aim for 6 or less options. For longer sets of options, consider using a [select](/components/Select) or an [action menu](/components/ActionMenu)
+- Radio buttons cannot be unchecked, so only use a radio group when a selection is required. Consider the following alternatives:
+  - use a radio group, but with one option that equates to "none of these options"
+  - use a [checkbox group](/components/CheckboxGroup) that fails validation if more than one option is selected
+- If there is an option that's the most likely option to be checked, that option should be checked by default
+
+## Accessibility
+
+- Once a user checks an option from a radio group, the result should not be saved or applied until the user has explicitly submitted their input using a "save" or "submit" button. A radio group is a single tab stop, and the options can be navigated using the arrow keys: this means assistive technologies like a screen reader cannot "read" an option without selecting it. See the [saving pattern guidelines](/ui-patterns/saving) for more information.
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:44]:
```suggestion
- Once a user checks an option from a radio group, the result should not be saved or applied until the user has explicitly submitted their input using a "save" or "submit" button. A radio group is a single tab stop, and the options can be navigated using the arrow keys: this means assistive technologies like a screen reader cannot "read" an option without selecting it. See the [saving pattern guidelines](https://primer.style/ui-patterns/saving) for more information.
```

File: apps/docs/content/components/RadioGroup/index.mdx [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+- Radio groups must be labeled. The label may be visually hidden, but visible labels are preferred.
+
+## Related components
+
+- [FormControl](/components/FormControl): Used to group form controls with labels
+- [Radio](/components/Radio): Individual radio button control
+- [CheckboxGroup](/components/CheckboxGroup): For allowing multiple selections
+- [SegmentedControl](/components/SegmentedControl): For immediate selection outside forms
```
Comment (rezrah) [inline apps/docs/content/components/RadioGroup/index.mdx:52]: This is a product component so let's remove it. Currently 404's too.

```suggestion
```

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.examples.stories.tsx [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+type Story = StoryObj<MetaProps>
+
+export const Inline: Story = {
+  args: {
+    labelChildren: 'Inline example',
+    labelVisuallyHidden: false,
+    captionChildren: 'Checkboxes can be displayed inline too',
+    validationChildren: '',
+  },
+  render: ({labelChildren, labelVisuallyHidden, captionChildren, validationChildren, validationVariant}) => {
```
Comment (rezrah) [inline packages/react/src/forms/CheckboxGroup/CheckboxGroup.examples.stories.tsx:20]: 🚀  This is cool...

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (rezrah) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:69]: Yeah, just feels redundant for us to have to declare this in every story file going forward 😕, maybe we make it a global type declaration?

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css [hunk 29]
```diff
@@ -0,0 +1,55 @@
...
+  top: -1px;
+}
+
+.ControlGroup__validation-icon--error {
+  position: relative;
+  top: -0.5px;
+}
+
+.ControlGroup__validation--animate-in {
+  animation: 170ms ControlGroupValidationFadeIn cubic-bezier(0.44, 0.74, 0.36, 1);
```
Comment (rezrah) [inline packages/react/src/forms/ControlGroup/ControlGroup.module.css:43]: Non-blocking: 

Appreciate these are probably from FormControl. We have newer animation-specific variables now, but none that match directly to this.

@joshfarrant, how close are these values to what's above? Can we sub them out? Can also do this in a separate PR.

File: packages/react/src/forms/FormControl/FormControl.module.css [hunk 30]
```diff
@@ -60,6 +60,10 @@
   width: 100%;
 }
 
+:global(.ControlGroup) .FormControl-label {
```
Comment (rezrah) [inline packages/react/src/forms/FormControl/FormControl.module.css:63]: I don't think this will work as expected in the compiled output, as .ControlGroup will be hashed in that components stylesheet but not here. Is there another way we do this from inside the ControlGroup component with higher specificity selector instead?

File: packages/react/src/forms/RadioGroup/RadioGroup.test.tsx [hunk 31]
```diff
@@ -0,0 +1,153 @@
...
+      </RadioGroup>,
+    )
+    const results = await axe(container)
+
+    expect(results).toHaveNoViolations()
+  })
+
+  describe('aria-describedby', () => {
```
Comment (rezrah) [inline packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:70]: Nit and more of a style preference: do we need the nested describe block? Could we instead rename the title of each test case to be more descriptive?

File: packages/react/src/forms/FormControl/FormControl.module.css (hunk 30 above)
Comment (joshfarrant) [reply packages/react/src/forms/FormControl/FormControl.module.css:63]: Resolved as per our conversation

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (joshfarrant) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:69]: `MetaProps` is unique to every component though (not the same as Storybook's `Meta` type), so any global type that we create would have the same signature as this built-in `StoryObj` type.

```ts
// If we created our own type, it would look like this
type CustomStoryObj<T> = StoryObj<T>

// So we'd replace
type Story = StoryObj<MetaProps>

// with
type Story = CustomStoryObj<MetaProps>
```

which doesn't really have any benefit.

File: .changeset/popular-seas-yawn.md [hunk 32]
```diff
@@ -0,0 +1,5 @@
+---
+'@primer/react-brand': minor
+---
+
+Added new `CheckboxGroup` and `RadioGroup` components to group `Checkbox` and `Radio` components respectively.
```
Comment (rezrah) [inline .changeset/popular-seas-yawn.md:5]: Sorry forgot to mention that for new components we usually add a bit more detail:

Something like:

( I can't make this a suggestion due to the code block syntax )

-- 


New `CheckboxGroup` and `RadioGroup` components to group `Checkbox` and `Radio` components are now available.

`CheckboxGroup` 

```jsx
<CheckboxGroup>
  <CheckboxGroup.Label>Choose your favorite features</CheckboxGroup.Label>
  <FormControl>
    <FormControl.Label>Actions notifications</FormControl.Label>
    <Checkbox value="actions" />
  </FormControl>
  <FormControl>
    <FormControl.Label>Packages</FormControl.Label>
    <Checkbox value="packages" />
  </FormControl>
  <FormControl>
    <FormControl.Label>Codespaces</FormControl.Label>
    <Checkbox value="codespaces" />
  </FormControl>
</CheckboxGroup>
```

:link: [Read the documentation for more `CheckboxGroup` examples](https://primer.style/brand/components/RadioGroup/react)


`RadioGroup`:

```jsx
<RadioGroup>
  <RadioGroup.Label>Choose your primary workspace</RadioGroup.Label>
  <FormControl>
    <FormControl.Label>Codespaces</FormControl.Label>
    <Radio name="workspace" value="codespaces" />
  </FormControl>
  <FormControl>
    <FormControl.Label>Local environment</FormControl.Label>
    <Radio name="workspace" value="local" />
  </FormControl>
  <FormControl>
    <FormControl.Label>Pen and paper</FormControl.Label>
    <Radio name="workspace" value="remote" />
  </FormControl>
</RadioGroup>
```

:link: [Read the documentation for more `RadioGroup` examples](https://primer.style/brand/components/RadioGroup/react)

**Review by rezrah — APPROVED** [review]
Thanks @joshfarrant, these new components are great ✨

File: packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (rezrah) [reply packages/react/src/forms/CheckboxGroup/CheckboxGroup.stories.tsx:69]: > not the same as Storybook's Meta type

Ohhh now I get it.. I can't read 🤦 ... ignore me. This is totally fine 😄

File: packages/react/src/forms/ControlGroup/ControlGroup.module.css (hunk 29 above)
Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:43]: Yes this was pulled from FormControl. You can see the three curves below. 

Our curve:
image

vs default:
image

vs glide:
image

I'd say that glide is probably the closest. Happy to switch this out

Comment (joshfarrant) [reply packages/react/src/forms/ControlGroup/ControlGroup.module.css:43]: I've just gone ahead and updated it to `brand-animation-duration-glide`

File: packages/react/src/forms/RadioGroup/RadioGroup.test.tsx (hunk 31 above)
Comment (joshfarrant) [reply packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:70]: Not required at all, but I quite like it as it

- Keeps test names more focussed
- Logically groups related tests in the code
- Groups tests in the output

image

Happy to remove it if you don't like it though

Comment (joshfarrant) [reply packages/react/src/forms/RadioGroup/RadioGroup.test.tsx:70]: Cool, updated 👍
